lib/polyfill/pip.js

```diff
@@ -0,0 +1,230 @@
+        this.addEventListener('webkitpresentationmodechanged',
+          proxyLeavePiPEvent);
+      } else {
+        // fallback for all the other events
+        oldAddEventListener.apply(this, [type, listener, options]);
+      }
+    };
+
+    const oldRemoveEventListener = proto.removeEventListener;
+    /**
+     * @this {HTMLMediaElement}
+     */
+    proto.removeEventListener = function(type, listener, options) {
+      if (type === 'enterpictureinpicture') {
+        const proxyEnterPiPEvent =
+          shaka.polyfill.PiP.getProxyEnterPiPEvent_(listener);
+        this.removeEventListener('webkitpresentationmodechanged',
+          proxyEnterPiPEvent);
+      } else if (type === 'leavepictureinpicture') {
+        const proxyLeavePiPEvent =
+          shaka.polyfill.PiP.getProxyLeavePiPEvent_(listener);
+        this.removeEventListener('webkitpresentationmodechanged',
+          proxyLeavePiPEvent);
+      } else {
+        // fallback for all the other events
+        oldRemoveEventListener.apply(this, [type, listener, options]);
+      }
+    };
+  }
+};
+
+/**
+ * @this {HTMLMediaElement}
+ * @private
+ */
+shaka.polyfill.PiP.updatePictureInPictureElementInDocument_ = function() {
+  if (this.webkitPresentationMode &&
+    this.webkitPresentationMode !== 'picture-in-picture') {
+    document.pictureInPictureElement = null;
+  }
+};
+
+/**
+ * @this {HTMLMediaElement}
+ * @return {!Promise}
+ * @private
+ */
+shaka.polyfill.PiP.requestPictureInPicture_ = function() {
+  // check if PIP is enabled
+  if (!this.webkitSupportsPresentationMode('picture-in-picture')) {
+    const error = new Error('PIP not allowed by videoElement',
+      'InvalidStateError');
+    return Promise.reject(error);
+  } else {
+    // enter PIP mode
+    this.webkitSetPresentationMode('picture-in-picture');
+    document.pictureInPictureElement = this;
+    return Promise.resolve();
+  }
+};
+
+/**
+ * @this {Document}
+ * @return {!Promise}
+ * @private
+ */
+shaka.polyfill.PiP.exitPictureInPicture_ = function() {
+  if (document.pictureInPictureElement) {
+    // exit PIP mode
+    const video =
+      /** @type {!HTMLMediaElement} */ (document.pictureInPictureElement);
+    video.webkitSetPresentationMode('inline');
+    document.pictureInPictureElement = null;
+    return Promise.resolve();
+  } else {
+    const error = new Error('No picture in picture element found',
+      'InvalidStateError');
+    return Promise.reject(error);
+  }
+};
+
+/**
+ * Get the proxy webkitpresentationmodechanged event.
+ *
+ * @param {EventListener} listener
+ * @return {!Function}
+ * @private
+ */
+shaka.polyfill.PiP.getProxyEnterPiPEvent_ = function(listener) {
+  return function(event) {
+    const videoElement = /** @type {HTMLVideoElement} */ (event.target);
+    if (videoElement.webkitPresentationMode === 'picture-in-picture') {
+      // keep track of the pipElement
+      document.pictureInPictureElement = videoElement;
+      listener();
```

Comment:
Here you'll need to call listener(event) with the event, since the compiler expects an argument.  But if the listener is looking at event.type for some reason, it might get confused by seeing a type it didn't listen for.

I think this is another good argument for redispatching events.